core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -333,6 +346,142 @@ public long nextFetchOffset() {
         }
     }
 
+    void updateOffsetsOnLsoMovement() {
+        long logStartOffset = offsetForEarliestTimestamp();
+        lock.writeLock().lock();
+        try {
+            if (logStartOffset <= startOffset) {
+                log.error("The log start offset: {} is not greater than the start offset: {} for the share partition: {}-{}",
+                        logStartOffset, startOffset, groupId, topicIdPartition);
+                return;
+            }
+            log.debug("Updating start offset for share partition: {}-{} from: {} to: {} since LSO has moved to: {}",
+                    groupId, topicIdPartition, startOffset, logStartOffset, logStartOffset);
+            if (cachedState.isEmpty()) {
+                // If the cached state is empty, then the start and end offset will be the new log start offset.
+                // This can occur during the initialization of share partition if LSO has moved.
+                startOffset = logStartOffset;
+                endOffset = logStartOffset;
+                return;
+            }
+            // Archive the available records in the cached state that are before the new log start offset.
+            boolean anyRecordArchived = archiveAvailableRecordsOnLsoMovement(logStartOffset);
+            // If we have transitioned the state of any batch/offset from AVAILABLE to ARCHIVED,
+            // then there is a chance that the next fetch offset can change.
+            if (anyRecordArchived) {
+                findNextFetchOffset.set(true);
+            }
```

Comment:
Shouldn't we check if new LSO is ahead of next fetch offset then only we need to re-compute else why to compute again?

Reply:
That condition is automatically incorporated. The function `updateOffsetsOnLsoMovement` is called only when realistically the nextFetchOffset is smaller than the LSO. Up until this point of the code, we are not changing the startOffset or endOffset, so there is no change to that condition.